tests/keygen.c

```diff
@@ -0,0 +1,206 @@
+		retval = -6;
+		goto err;
+	}
+	if(1 != EVP_DigestSignFinal(mdctx, sig, &slen)) {
+		display_openssl_errors(__LINE__);
+		retval = -7;
+		fprintf(stderr, "Sign fail\n");
+		goto err;
+	}
+	printf("Sign success\n");
+
+	if(1 != EVP_DigestVerifyInit(mdctx, NULL, EVP_sha256(), NULL, pub)) {
+		display_openssl_errors(__LINE__);
+		retval = -8;
+		goto err;
+	}
+	if(1 != EVP_DigestVerifyUpdate(mdctx, msg, strlen(msg))) {
+		display_openssl_errors(__LINE__);
+		retval = -9;
+		goto err;
+	}
+	if(1 == EVP_DigestVerifyFinal(mdctx, sig, slen))
+	{
+		printf("Verify success\n");
+		retval = 0;
+	}
+	else
+	{
+		display_openssl_errors(__LINE__);
+		fprintf(stderr, "Verify fail\n");
+		retval = -10;
+		goto err;
+	}
+
+err:
+	if(sig) OPENSSL_free(sig);
+	if(mdctx) EVP_MD_CTX_destroy(mdctx);
+	return retval;
+}
+
+int
+main(int argc, char* argv[])
+{
+	int ret = 0;
+	ENGINE* engine = NULL;
+	char *module = argv[3];
+
+	if (argc < 3) {
+		fprintf(stderr, "Too few arguments\n");
+		usage(argv);
+		return 1;
+	}
+
+	ENGINE_load_builtin_engines();
+	engine = ENGINE_by_id("pkcs11");
+	if (engine == NULL) {
+		fprintf(stderr, "Could not get engine\n");
+		display_openssl_errors(__LINE__);
+		exit(1);
+	}
+	if (!ENGINE_ctrl_cmd_string(engine, "PIN", "1234", 0)) {
+		display_openssl_errors(__LINE__);
+		exit(1);
+	}
+	if (!ENGINE_ctrl_cmd_string(engine, "VERBOSE", NULL, 0)) {
+		display_openssl_errors(__LINE__);
+		exit(1);
+	}
+	if (module) {
+		if (!ENGINE_ctrl_cmd_string(engine, "MODULE_PATH", module, 0)) {
+			display_openssl_errors(__LINE__);
+			exit(1);
+		}
+	}
+	if (!ENGINE_init(engine)) {
+		fprintf(stderr, "Could not initialize engine\n");
+		display_openssl_errors(__LINE__);
+		exit(1);
+	}
+
+	PKCS11_EC_KGEN ec = {
+		.curve = "P-256"
+	};
+	PKCS11_KGEN_ATTRS eckg =
+	{
+		.type = EVP_PKEY_EC,
+		.kgen.ec = &ec,
+		.token_label = argv[1],
+		.key_label = argv[2],
+		.key_id = "1234",
+	};
+
+	if (!ENGINE_ctrl_cmd(engine, "KEYGEN", 0, &eckg, NULL, 1)) {
+		fprintf(stderr, "Could not generate ECC keys\n");
+		exit(1);
+	}
+
+	EVP_PKEY *ecpb = ENGINE_load_public_key(engine, "1234", NULL, NULL);
+	EVP_PKEY *ecpr = ENGINE_load_private_key(engine, "1234", NULL, NULL);
+	if ((ret = sign_verify_test(ecpr, ecpb)) < 0) {
+		fprintf(stderr, "ECC Sign-verify failed with err code: %d\n", ret);
+		exit(1);
+	}
+	PKCS11_RSA_KGEN rsa = {
+		.bits = 2048
+	};
+	PKCS11_KGEN_ATTRS rsakg = {
+		.type = EVP_PKEY_RSA,
+		.kgen.rsa = &rsa,
+		.token_label = argv[1],
+		.key_label = argv[2],
+		.key_id = "4321",
+	};
+	if (!ENGINE_ctrl_cmd(engine, "KEYGEN", 0, &rsakg, NULL, 1)) {
+		fprintf(stderr, "Could not generate RSA keys\n");
+		exit(1);
+	}
+	EVP_PKEY *rsapb = ENGINE_load_public_key(engine, "4321", NULL, NULL);
+	EVP_PKEY *rsapr = ENGINE_load_private_key(engine, "4321", NULL, NULL);
+	if ((ret = sign_verify_test(rsapr, rsapb)) < 0) {
+		fprintf(stderr, "RSA Sign-verify failed with err code: %d\n", ret);
+		exit(1);
+	}
+
+	ENGINE_finish(engine);
```

Comment:
Can you also add test cases that extract the respective keys and use those to verify the signature?

Reply:
This is what the test does. After successful ENGINE_ctrl_cmd(engine, "KEYGEN", 0, &eckg, NULL, 1); the keypair is living on the HSM. Then we extract it with 

EVP_PKEY *ecpb = ENGINE_load_public_key(engine, "1234", NULL, NULL);
EVP_PKEY *ecpr = ENGINE_load_private_key(engine, "1234", NULL, NULL);

and use it to sign and verify.